iOS migration guide. (#341)
pipecat-ai/docs: client/ios/migration-guide.mdx

---

title: "RTVIClient Migration Guide for IOS"
sidebarTitle: "Migration Guide"
description: "A Guide to migrating from RTVIClient to PipecatClient for iOS"

> ----------------------------------------------------------------------------

Reviewer: ```suggestion
```

**Old**

```swift
let client = RTVIClient(params: ...)
```

**New**

```swift
let pipecatClientOptions = PipecatClientOptions(
    transport: DailyTransport(),
    enableMic: true,
    enableCam: false
)

let client = PipecatClient(options: pipecatClientOptions)
```

### 3. Connection Method

The connection process is now explicit and supports modern async handling.

**Old**

```swift
client.connect()
```

**New**

```swift
let startBotParams = APIRequest(endpoint: URL(string: "https://your-server/connect")!)
client.startBotAndConnect(startBotParams: startBotParams) { result in
    switch result {
    case .success(let connectionParams):
        print("Connected successfully: \(connectionParams)")
    case .failure(let error):
        print("Connection failed: \(error)")
    }
}
```

### 4. Function Call Handling

Helpers like `LLMHelper` have been removed. You now register function call handlers directly on the `PipecatClient`.

**Old**

```swift
let llmHelper = LLMHelper()

> client.registerHelper("llm", llmHelper)

Reviewer 2: can this be expanded to show whatever was previously required to specifically handle a function call? I want to make it clear that `registerFunctionCallHandler` only replaces that specific functionality and not all of what `llmHelper` could do before.

Author: Done.

```

**New**

```swift
client.registerFunctionCallHandler(functionName: "get_my_current_location") { functionCallData, onResult in
    let location = await getCurrentLocation()
    await onResult(location)
}

client.registerFunctionCallHandler(functionName: "set_restaurant_location") { functionCallData, onResult in
    handleRestaurantLocation(functionCallData.args)
    await onResult(.string("success"))
}
```

### 5. Pipeline Configuration Initialization

> Previously, you could provide a pipeline configuration as part of the `RTVIClient` constructor and it was expected to be in a specific format. Now, if you would like to pass any initial pipeline configurations, you do so as `requestData` added to the endpoint you provide to `connect()`. In both cases, you would need server-side code to parse and apply these settings, but now you can define the structure and what pieces of configuration you want to send.

Reviewer 2: ```suggestion
Previously, you could provide a pipeline configuration as part of the `RTVIClient` constructor and it was expected to be in a specific format. Now, if you would like to pass any initial pipeline configurations, you do so as `requestData` added to the endpoint you provide to `startBot()` or `startBotAndConnect()`. In both cases, you would need server-side code to parse and apply these settings, but now you can define the structure and what pieces of configuration you want to send.
```

I need to update the js migration too

Check out [this section of docs](../js/api-reference/messages#connection-time-configuration) for an example, complete with server-side code showing how to initialize the pipeline configuration at connection time.

### 6. LLM Context Updates

Previously, context updates were done via helpers. Now, use `appendToContext()`:

```swift
client.appendToContext(
    role: "user",
    content: "Tell me a joke.",
    run_immediately: true
)
```

> ### 7. Sending Messages and Requests

Reviewer 2: Is there a reason you changed this title from "PipelineConfiguration Updates"?

Author: Done.

> Previously, the client supported updating the pipeline configuration using a specific method that took a configuration object in a generic format. Dynamic and predifined configuration updates, however, are a security concern, allowing clients to override settings and potentially abuse API keys. For this reason, it has been removed and most configuration updates need to be handled custom by your application. To do so, you should take advantage of the client-server messaging system, which allows you to send messages to the server and handle responses. This way, you can implement your own logic for updating configurations securely.

Reviewer 2: ```suggestion
Previously, the client supported updating the pipeline configuration using a specific method that took a configuration object in a generic format. Dynamic and predefined configuration updates, however, are a security concern, allowing clients to override settings and potentially abuse API keys. For this reason, it has been removed and most configuration updates need to be handled custom by your application. To do so, you should take advantage of the client-server messaging system, which allows you to send messages to the server and handle responses. This way, you can implement your own logic for updating configurations securely.
```

New messaging types replace old action/config helpers:

```swift
try client.sendClientMessage(msgType: "set-voice", data: .string("Janice"))

let llmVendor = try await client.sendClientRequest(msgType: "get-llm-vendor").d?.asString
print("LLM Vendor: \(llmVendor ?? "")")
```

Check out [this section of docs](../js/api-reference/messages#sending-custom-messages-to-the-server) for a more complete example, along with an example on making a request (`sendClientRequest()`) to wait for a response.

> ### 7. Disconnecting

Reviewer: ```suggestion
### 8. Disconnecting
```

1. **Constructor Params Removed**: No pipeline or endpoint config in constructor.
2. **Helpers Removed**: RTVIClientHelper, LLMHelper are gone.
3. **Configs Removed**: All configuration-related methods, events, and types have been removed: `getConfig()`, `updateConfig()`, `describeConfig()`, `onConfigUpdated`, `onConfigDescribed`, etc.
4. **Actions Removed**: All actions-related methods, events, and types have been removed: `action()`, `describeActions()`, `onActionsAvailable`, etc.
8. **New Messaging Methods**:

   * `appendToContext()`
   * `sendClientRequest()`
   * `sendClientMessage()`
   * `disconnectBot()`
   * `registerFunctionCallHandler()` / `unregisterFunctionCallHandler()` / `unregisterAllFunctionCallHandlers()`

## Migration Steps

1. Update imports to `PipecatClientIOS` and `PipecatClientIOSDaily`.
2. Move connection configuration from constructor to `startBotAndConnect()` or `connect()` methods.
3. Replace helper-based function calls with `registerFunctionCallHandler()`.
4. Replace context updates with `appendToContext()`.
5. Replace all action and config related methods, events, and types with `sendClientMessage()` or `sendClientRequest()` for custom messages.

> 6. Ensure proper disconnection via `unregisterAllFunctionCallHandlers()` and `disconnect()`.

Reviewer 2: did this change?
